### PR TITLE
Provide Apollo dynamic configuration center implementation. fixes #2825

### LINKED ISSUE #2825: Dynamic Configuration Apollo Support
Please answer these questions before submitting your issue.

- Why do you submit this issue?
- [ ] Question or discussion
- [ ] Bug
- [ ] Requirement
- [x] Feature or performance improvement

___
### Requirement or improvement

related issue: #2816 #2810 

[Apollo](https://github.com/ctripcorp/apollo) is another popular configuration center, we'd better support it as well, I'm familiar with Apollo and I'd like to contribute on this feature :-)

### PR DESCRIPTION
**Pull request by kezhenxu94**
Please answer these questions before submitting pull request

- Why submit this pull request?
- [ ] Bug fix
- [x] New feature provided
- [ ] Improve performance

- Related issues
#2825

### CODE REVIEW

**Comment by kezhenxu94** [comment]
checking dependency licenses

**Comment by kezhenxu94** [comment]
@wu-sheng Here are the differencies of dependencies between the dist package of my PR and the official dist 6.1.0 package (downloaded from the official website)

image

all the newly added dependencies are Apache 2.0 licenses, attachments are the dependencies names:

[swold.txt](https://github.com/apache/skywalking/files/3255750/swold.txt)
[swnew.txt](https://github.com/apache/skywalking/files/3255751/swnew.txt)


- apollo-client
- apollo-core
- builder-annotations
- commons-compress
- fastjson
- guice-4.1.0.jar 
- javax.inject-1.jar 
- joda-convert-1.2.jar 
- resourcecify-annotations-0.9.2.jar 
- sundr-codegen-0.9.2.jar 
- sundr-core-0.9.2.jar

**Comment by wu-sheng** [comment]
I will follow this next week. Many new libs.

**Comment by kezhenxu94** [comment]
I've updated the newly-added dependencies list in my original comment, there're fewer **because last time I compared the master branch with the latest released version**, and there're some dependencies that are introduced by other commits (e.g. dependency `io.sundr:sundr-core:jar:0.9.2` is introduced by [this PR](https://github.com/apache/skywalking/pull/2460/files#diff-7c61192cd2fd51bddd8930cc37842523R42) as a transitive dependency of kubernetes):

image

**Comment by wu-sheng** [comment]
I am back online. For this pull request, please read this PR #2640 .  Your #2817 is easy to merge because of that. 

All dependency libs must be added to LICENSE file in dist folder. For Apache 2.0, NOTICE file content should be added to our NOTICE. Other licenses should be copied. These include the `io.sundr:sundr-core:jar:0.9.2`, if we missed that when we updated the k8s client version.

**Comment by kezhenxu94** [comment]
> These include the `io.sundr:sundr-core:jar:0.9.2`, if we missed that when we updated the k8s client version.

here is the differences between this PR (based on latest master branch) and the official release 6.1.0, and I'll base on the diff to add missing license or notice:

```diff
diff --git a/tmp/old.txt b/tmp/new.txt
index b814f5ce9..0d67fa4a7 100644
--- a/tmp/old.txt
+++ b/tmp/new.txt
@@ -4,25 +4,34 @@ aggs-matrix-stats-client-6.3.2.jar
 animal-sniffer-annotations-1.17.jar
 annotations-13.0.jar
 antlr4-runtime-4.7.1.jar
+aopalliance-1.0.jar
 apm-datacarrier-6.1.0.jar
 apm-network-6.1.0.jar
 apm-util-6.1.0.jar
+apollo-client-1.4.0.jar
+apollo-core-1.4.0.jar
 bcpkix-jdk15on-1.59.jar
 bcprov-ext-jdk15on-1.59.jar
 bcprov-jdk15on-1.59.jar
+builder-annotations-0.9.2.jar
 caffeine-2.6.2.jar
-client-java-2.0.0.jar
-client-java-api-2.0.0.jar
-client-java-proto-2.0.0.jar
+client-java-4.0.0.jar
+client-java-api-4.0.0.jar
+client-java-proto-4.0.0.jar
 cluster-consul-plugin-6.1.0.jar
 cluster-kubernetes-plugin-6.1.0.jar
+cluster-nacos-plugin-6.1.0.jar
 cluster-standalone-plugin-6.1.0.jar
 cluster-zookeeper-plugin-6.1.0.jar
 commons-codec-1.11.jar
+commons-compress-1.18.jar
 commons-dbcp-1.4.jar
 commons-io-2.6.jar
 commons-lang3-3.7.jar
 commons-pool-1.5.4.jar
+configuration-api-6.1.0.jar
+configuration-apollo-6.1.0.jar
+configuration-nacos-6.1.0.jar
 consul-client-1.2.6.jar
 converter-jackson-2.3.0.jar
 curator-client-4.0.1.jar
@@ -39,10 +48,12 @@ elasticsearch-x-content-6.3.2.jar
 envoy-metrics-receiver-plugin-6.1.0.jar
 error_prone_annotations-2.2.0.jar
 exporter-6.1.0.jar
+fastjson-1.2.47.jar
 generated-analysis-6.1.0.jar
 graphql-java-8.0.jar
 graphql-java-tools-5.2.3.jar
 groovy-2.4.5-indy.jar
+grpc-configuration-sync-6.1.0.jar
 grpc-context-1.15.1.jar
 grpc-core-1.15.1.jar
 grpc-netty-1.15.1.jar
@@ -51,6 +62,7 @@ grpc-protobuf-lite-1.15.1.jar
 grpc-stub-1.15.1.jar
 gson-2.8.1.jar
 guava-20.0.jar
+guice-4.1.0.jar
 h2-1.4.196.jar
 hppc-0.7.1.jar
 httpasyncclient-4.1.2.jar
@@ -70,6 +82,7 @@ jackson-mapper-asl-1.9.13.jar
 jackson-module-kotlin-2.8.8.jar
 jaeger-receiver-plugin-6.1.0.jar
 java-dataloader-2.0.2.jar
+javax.inject-1.jar
 javax.servlet-api-3.1.0.jar
 jcl-over-slf4j-1.7.25.jar
 jetty-http-9.4.2.v20170220.jar
@@ -80,6 +93,7 @@ jetty-servlet-9.4.2.v20170220.jar
 jetty-util-9.4.2.v20170220.jar
 jline-0.9.94.jar
 jna-4.5.1.jar
+joda-convert-1.2.jar
 joda-time-2.9.9.jar
 jopt-simple-4.6.jar
 jsr305-1.3.9.jar
@@ -111,6 +125,9 @@ lucene-spatial-7.3.1.jar
 lucene-spatial-extras-7.3.1.jar
 lucene-spatial3d-7.3.1.jar
 lucene-suggest-7.3.1.jar
+nacos-api-1.0.0.jar
+nacos-client-1.0.0.jar
+nacos-common-1.0.0.jar
 netty-3.10.5.Final.jar
 netty-buffer-4.1.27.Final.jar
 netty-codec-4.1.27.Final.jar
@@ -137,6 +154,7 @@ rank-eval-client-6.3.2.jar
 reactive-streams-1.0.2.jar
 receiver-proto-6.1.0.jar
 reflectasm-1.11.7.jar
+resourcecify-annotations-0.9.2.jar
 retrofit-2.3.0.jar
 server-alarm-plugin-6.1.0.jar
 server-core-6.1.0.jar
@@ -159,6 +177,8 @@ storage-elasticsearch-plugin-6.1.0.jar
 storage-jaeger-plugin-6.1.0.jar
 storage-jdbc-hikaricp-plugin-6.1.0.jar
 storage-zipkin-plugin-6.1.0.jar
+sundr-codegen-0.9.2.jar
+sundr-core-0.9.2.jar
 swagger-annotations-1.5.12.jar
 t-digest-3.2.jar
 telemetry-api-6.1.0.jar
                                                                              
```

**Comment by kezhenxu94** [comment]
HI @wu-sheng 

- [aopalliance](http://aopalliance.sourceforge.net)

> LICENCE: all the source code provided by AOP Alliance is Public Domain.

**PUBLIC & NO NOTICE FILE** (`rsync -a a.cvs.sourceforge.net::cvsroot/aopalliance/`)

- apollo-client-1.4.0.jar and apollo-core-1.4.0.jar: **Apache 2.0 & NO LICENSE FILE**

- builder-annotations-0.9.2.jar resourcecify-annotations-0.9.2.jar sundr-codegen-0.9.2.jar
sundr-core-0.9.2.jar (https://github.com/sundrio/sundrio): **Apache 2.0 & NO NOTICE FILE**

- commons-compress-1.18.jar: **Apache 2.0 & NOTICE file contents are copied**

- fastjson-1.2.47.jar (**EXISTS in 6.1.0 and LICENSE is added**)

- guice-4.1.0.jar: **Apache 2.0 & NO NOTICE**

- javax.inject-1.jar: **Apache 2.0 & NO NOTICE**

- joda-convert-1.2.jar: **Apache 2.0 & NOTICE file contents are copied**

- nacos-api-1.0.0.jar nacos-client-1.0.0.jar nacos-common-1.0.0.jar: **Apache 2.0 & NO NOTICE file**

**Comment by wu-sheng** [comment]
You need to cross check the difference between libs and existing license file.

**Comment by kezhenxu94** [comment]
> You need to cross check the difference between libs and existing license file.

Do you mean version or anything else?

**Comment by wu-sheng** [comment]
@kezhenxu94 As we have known, we can't do fully CI tests in travis. I suggest that, we set up the jenkins tests first, then we discuss how to merge this.

**Comment by kezhenxu94** [comment]
> @kezhenxu94 As we have known, we can't do fully CI tests in travis. I suggest that, we set up the jenkins tests first, then we discuss how to merge this.

Jenkins tests are running now and I've dealt with the new dependencies' licenses, so I suggest reconsidering this PR, what do you think, @wu-sheng @JaredTan95  :)

**Comment by kezhenxu94** [comment]
Some of the dependencies listed above come from another PR #2460 and I've added in another PR #2879

**Comment by wu-sheng** [comment]
Jenkins test fail, could you recheck?

**Comment by wu-sheng** [comment]
@JaredTan95 Please review.

**Comment by kezhenxu94** [comment]
> @JaredTan95 Please review.

I've tested enough, I know @JaredTan95 you are an Apollo contributor too :) , it would be appreciated if you could review this PR, thanks in advance 😄

File: docs/en/setup/backend/dynamic-config.md [hunk 1]
```diff
@@ -31,6 +31,20 @@ configuration:
     #clusterName: "default" # the name of current cluster, set the name if you want to upstream system known.  
 ```
 
+## Apollo DCS
```
Comment (wu-sheng) [inline docs/en/setup/backend/dynamic-config.md]: Please remove `DCS`. DCS is a gRPC service protocol defined in SkyWalking. Read the current document. :) I think you mean `Dynamic Configuration Apollo Implementation`

File: oap-server/server-configuration/configuration-apollo/src/test/resources/docker/docker-entrypoint-initdb.d/apolloconfigdb.sql [hunk 2]
```diff
@@ -0,0 +1,539 @@
...
+  KEY `DataChange_LastTime` (`DataChange_LastTime`)
+) ENGINE=InnoDB AUTO_INCREMENT=6 DEFAULT CHARSET=utf8mb4;
+/*!40101 SET character_set_client = @saved_cs_client */;
+
+--
+-- Dumping data for table `ServerConfig`
+--
+
+/*!40000 ALTER TABLE `ServerConfig` DISABLE KEYS */;
+INSERT INTO `ServerConfig` VALUES (1,'eureka.service.url','default','http://localhost:8080/eureka/','Eureka服务Url，多个service以英文逗号分隔','\0','default','2019-06-06 15:14:00','','2019-06-06 15:14:00'),(2,'namespace.lock.switch','default','false','一次发布只能有一个人修改开关','\0','default','2019-06-06 15:14:00','','2019-06-06 15:14:00'),(3,'item.value.length.limit','default','20000','item value最大长度限制','\0','default','2019-06-06 15:14:00','','2019-06-06 15:14:00'),(4,'config-service.cache.enabled','default','false','ConfigService是否开启缓存，开启后能提高性能，但是会增大内存消耗！','\0','default','2019-06-06 15:14:00','','2019-06-06 15:14:00'),(5,'item.key.length.limit','default','128','item key 最大长度限制','\0','default','2019-06-06 15:14:00','','2019-06-06 15:14:00');
```
Comment (wu-sheng) [inline oap-server/server-configuration/configuration-apollo/src/test/resources/docker/docker-entrypoint-initdb.d/apolloconfigdb.sql]: Make Chinese config, need to be removed or changed.

**Comment by wu-sheng** [comment]
CI fail because of new changes.

**Pull request merged**